Change "Open cohort in Data Explorer" to new way of opening tools

melissachang opened the pull request:
FYI @seanhorgan @jeromechadel 

Before:

![1](https://user-images.githubusercontent.com/10929390/63202177-5dc15480-c03d-11e9-88e8-2194a49f8890.png)

After:

Don't show "Data Explorer" for entities that aren't cohorts:

<img width="1552" alt="1" src="https://user-images.githubusercontent.com/10929390/63217728-747caf80-c100-11e9-8ad2-1e79a9681949.png">

For cohorts, only show "Data Explorer" and don't show IGV or Workflow. 1 cohort selected:

![1](https://user-images.githubusercontent.com/10929390/63464577-65557480-c414-11e9-8473-9b9817910800.png)

2 cohorts selected:

![2](https://user-images.githubusercontent.com/10929390/63464584-68e8fb80-c414-11e9-96de-b385a6f4c140.png)

(I disabled the button in this case but you can't tell from the looks of it.)

Also fix this harmless but confusing console error by wrapping `iframeResizer` in `onLoad()`.

```
Failed to execute 'postMessage' on 'DOMWindow': The target origin provided ('https://test-data-explorer.appspot.com') does not match the recipient window's origin ('http://localhost:3000').
```

melissachang committed:
Change "Open cohort in Data Explorer" to new way of opening tools

lint

Don't show other tools for cohorts

use data explorer logo

Show all buttons for all entities

implement disabled tool button

optimize svg

final edits

zarsky-broad commented:
Closing for now, as there are apparently some changes in the pipeline for this.

melissachang commented:
Waiting for @jeromechadel to decide if we want to show all buttons for all entity types

seanhorgan commented:
Shouldn't we just display tools that are capable of opening the respective data type?

melissachang commented:
I believe @jeromechadel wants to show all buttons for discoverability. There aren't that many tools now; later on when there are more tools, this could change. I'd like to get something in because this unblocks other work. So I'll show all buttons, and this can be revisted later if necessary.

This PR is ready to be reviewed. Updated screenshots:

If non-cohort entity is select, button is disabled (though can't tell from the look) and tooltip says "Only cohorts can be opened with Data Explorer":

![1](https://user-images.githubusercontent.com/10929390/64447726-e9ecf780-d090-11e9-8957-3d6ae2c6c659.png)

If cohort is selected, IGV button is disabled and tooltip says "IGV cannot be opened with cohorts".

![1](https://user-images.githubusercontent.com/10929390/64448040-9202c080-d091-11e9-8cbe-60e7849b23bd.png)

If cohort is selected and doesn't have data_explorer_url, button is disabled and tooltip says "Cohort is too old, please recreate in Data Explorer and save to Terra again". data_explorer_url has been [populated since April 2019](https://github.com/DataBiosphere/data-explorer/pull/318), so this shouldn't affect any real Terra users.

![1](https://user-images.githubusercontent.com/10929390/64448628-fc683080-d092-11e9-8727-81e8f6a554dd.png)

If more than one cohort is selected, button is disbled and tooltip says "Select exactly one cohort to open in Data Explorer"

![1](https://user-images.githubusercontent.com/10929390/64448823-67196c00-d093-11e9-8c36-e446a6f28b16.png)

zarsky-broad commented on src/components/DataExplorerFrame.js:
```diff
@@ -9,32 +9,32 @@ import * as Nav from 'src/libs/nav'
+        }
+      }, elem.current)
+
+      return elem.current.iFrameResizer.removeListeners
```
Returning this from `useEffect` causes it to be run on unmount; however, doing so from here doesn't. You'll need to set this up to run on unmount.

melissachang replied:
Reverted changes to this file

zarsky-broad commented on src/images/data-explorer-logo.svg:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Please run `npm run optimize-image-svgs`

zarsky-broad reviewed (changes requested):
I'd want a visual disabled state from Lou for the buttons in order to do this.

melissachang commented:
This PR is ready to be reviewed again. Everything, including tooltips, has been approved by Jerome. Added visual disabled state.

I updated the screenshots in [this comment](https://github.com/DataBiosphere/terra-ui/pull/1805#issuecomment-528949497).

melissachang committed:
rebase

melissachang committed:
rebase fix

melissachang committed:
fix rebase

melissachang committed:
rename var

melissachang committed:
Fix console error when hovering over disabled tool button

If I select a cohort and hover over "IGV" button, console has:

Interactive.js:37 Assertion failed: border needs to be added to the hover-style in style.css for the style to be applied

melissachang committed:
fix workspace id

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -57,10 +58,10 @@ const styles = {
 
-export const ModalToolButton = ({ children, ...props }) => {
+export const ModalToolButton = ({ children, disabled, ...props }) => {
   return h(Clickable, _.merge({
```
Should we also pass disabled through here so the cursor doesn't change?

melissachang replied:
Done

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -358,10 +359,17 @@ const ReferenceDataContent = ({ workspace: { workspace: { namespace, attributes
+       entityKey === 'cohort' && _.size(selectedEntities) === 1 && _.values(selectedEntities)[0].attributes.data_explorer_url ?
+         _.values(selectedEntities)[0].attributes.data_explorer_url + '&wid=' + workspaceId :
+         ''
+  const dataExplorerButtonEnabled = dataExplorerUrl !== ''
```
Why not define this conditional as `dataExplorerButtonEnabled`, and then use it to set the url to either the value (I'd recommend using a string literal instead of concatenation) or undefined? Something like this:

```js
  const dataExplorerButtonEnabled = entityKey === 'cohort' && entitiesCount === 1 && !!_.values(selectedEntities)[0].attributes.data_explorer_url
  const dataExplorerUrl = dataExplorerButtonEnabled ? ${_.values(selectedEntities)[0].attributes.data_explorer_url}&wid=${workspaceId}` : undefined
```

I'd love to also find a way to dedupe the `_.values(selectedEntities)[0].attributes.data_explorer_url` part.

melissachang replied:
Done except for deduping. The way I had it is the only concise way to dedupe that I can think of.

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -383,23 +391,42 @@ const ToolDrawer = withModalDrawer()(({ workspace, onDismiss, onIgvSuccess, sele
             h(ModalToolButton, {
-              onClick: () => setToolMode('IGV'),
-              tooltip: 'Open with Integrative Genomics Viewer'
+              onClick: () => entityKey !== 'cohort' && setToolMode('IGV'),
```
If you define `const isCohort = entityKey === 'cohort'` somewhere above, you can avoid repeating this check each time.

Also, if this is disabled, you don't need to remove the onClick manually.

(These apply to the buttons below as well)

melissachang replied:
Done. I also got rid of `entitiesType` on line 367 since `entityKey` can be used instead.

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -383,23 +391,42 @@ const ToolDrawer = withModalDrawer()(({ workspace, onDismiss, onIgvSuccess, sele
               'Workflow'
+            ]),
+            h(ModalToolButton, {
+              onClick: () => dataExplorerButtonEnabled && window.open(dataExplorerUrl + '&wid=' + workspaceId),
```
The constructed url already has the workspace id in it.

melissachang replied:
Done

zarsky-broad reviewed (changes requested):
A few tweaks, and then we should be good to go.

melissachang committed:
review fixes

zarsky-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -423,7 +448,7 @@ const ToolDrawer = withModalDrawer()(({ workspace, onDismiss, onIgvSuccess, sele
       }
     }, [
-      `${entitiesCount} ${entitiesType + (entitiesCount > 1 ? 's' : '')} selected`
+      `${entitiesCount} ${entityKey + (entitiesCount > 1 ? 's' : '')} selected`
```
Was this an intentional change?

zarsky-broad replied:
Ah, nvm.

zarsky-broad reviewed (approved):
~One question, then~ 👍
